### ISSUE TITLE
RS w/ UnicastOnly=off & ClientList triggers RA flood

### ISSUE THREAD

**Issue by robbat2** [reubenhwk/radvd#63]
Expected result:
If a `UnicastOnly=no` instance with a fixed client list gets a RS, it should only respond to that client.

Actual result:
The instance sends RAs to every single client.

Effects:
Hosts that did not solicit the RA are flooded by RAs when not needed.

Host setup:
```
modprobe dummy
ifconfig dummy0 up multicast
ip -6 addr add  fe80::1/64 dev dummy0
ip -6 addr add  fe80::2/64 dev dummy0
```
radvd.conf:
```
interface dummy0 {
  AdvSendAdvert on;
  AdvRASrcAddress { fe80::1; };
  route 2001:db8:f001::/64 { };
  UnicastOnly off;
  clients { fe80::2; fe80::3; fe80::4; fe80::5; fe80::6; fe80::7; fe80::8; };
};
```

Test trigger:
```
# rdisc6 dummy0
Soliciting ff02::2 (ff02::2) on dummy0...

Hop limit                 :           64 (      0x40)
Stateful address conf.    :           No
Stateful other conf.      :           No
Mobile home agent         :           No
Router preference         :       medium
Neighbor discovery proxy  :           No
Router lifetime           :         1800 (0x00000708) seconds
Reachable time            :  unspecified (0x00000000)
Retransmit time           :  unspecified (0x00000000)
 Route                    : 2001:db8:f001::/64
  Route preference        :       medium
  Route lifetime          :         1800 (0x00000708) seconds
 Source link-layer address: 96:29:9A:3F:1A:A6
 from fe80::1
```

debug=5 logging
```
[Nov 18 17:43:12] radvd (16547): dummy0 recvmsg len=8
[Nov 18 17:43:12] radvd (16547): dummy0 received a packet
[Nov 18 17:43:12] radvd (16547): dummy0 received RS from: fe80::2
[Nov 18 17:43:12] radvd (16547): sending RA to fe80::8 on dummy0 (fe80::1)
[Nov 18 17:43:12] radvd (16547): sending RA to fe80::7 on dummy0 (fe80::1)
[Nov 18 17:43:12] radvd (16547): sending RA to fe80::6 on dummy0 (fe80::1)
[Nov 18 17:43:12] radvd (16547): sending RA to fe80::5 on dummy0 (fe80::1)
[Nov 18 17:43:12] radvd (16547): sending RA to fe80::4 on dummy0 (fe80::1)
[Nov 18 17:43:12] radvd (16547): sending RA to fe80::3 on dummy0 (fe80::1)
[Nov 18 17:43:12] radvd (16547): sending RA to fe80::2 on dummy0 (fe80::1)
[Nov 18 17:43:12] radvd (16547): dummy0 next scheduled RA in 496.716 second(s)
[Nov 18 17:43:12] radvd (16547): dummy0 processed an RS
[Nov 18 17:43:12] radvd (16547): polling for 496.716 second(s), next iface is dummy0
[Nov 18 17:43:12] radvd (16547): dummy0 recvmsg len=48
[Nov 18 17:43:12] radvd (16547): dummy0 received a packet
[Nov 18 17:43:12] radvd (16547): dummy0 received RA from: fe80::1
[Nov 18 17:43:12] radvd (16547): processed RA on dummy0
```

Network capture:
```
17:43:12.500909 IP6 fe80::2 > ff02::2: ICMP6, router solicitation, length 8
17:43:12.501121 IP6 fe80::1 > fe80::8: ICMP6, router advertisement, length 48
17:43:12.501170 IP6 fe80::1 > fe80::7: ICMP6, router advertisement, length 48
17:43:12.501201 IP6 fe80::1 > fe80::6: ICMP6, router advertisement, length 48
17:43:12.501233 IP6 fe80::1 > fe80::5: ICMP6, router advertisement, length 48
17:43:12.501265 IP6 fe80::1 > fe80::4: ICMP6, router advertisement, length 48
17:43:12.501313 IP6 fe80::1 > fe80::3: ICMP6, router advertisement, length 48
17:43:12.501343 IP6 fe80::1 > fe80::2: ICMP6, router advertisement, length 48
```

**Comment by robbat2** [comment]
If `UnicastOnly=on` is set, then the flood does not happen.

**Comment by reubenhwk** [comment]
Hmm..  I can picture the code, and that makes sense, and it should be fixed.  One of the problems is that the RA is randomly delayed, so once radvd gets an RS, it just schedules the RA to happen after a short delay, but doesn't remember the RS.  So the info in the RS needs to be linked to the RA somehow so it known to whom it should send the RA.  The schedule/delay queue will just need to store the source of the RS, *but* this is where it can get complicated.  Currently, only one RA is scheduled.  If 0, 1, or N RS's come in, still just one RA is scheduled.  This is fine because the RA is multicast to the all-nodes group.  So the RA timer queue will need to have one slot per client in the client list (I guess).

Kinda just thinking out loud here, I'll look into the code more later when I get time.  If you want to take a stab at fixing it, please do.  :)

**Comment by robbat2** [comment]
See RFC 7772 section 2.1: https://tools.ietf.org/html/rfc7772#section-2.1
In the case of a large network, this leads to a `all-nodes` RA every 3 seconds, which wakes up all of the devices on the network.

Suggested new behavior:
- when the process gets a RS, always send an immediate unicast response to only that client, with the complete set of RA options.
- Leave the `all-nodes` RA cycle alone (the new RA-splitting code has placeholders for scheduling RA options at different rates).

**Comment by reubenhwk** [comment]
That does sound like a better idea.  Just to clarify...

Leave all-node multicast cycle alone?  Unicast responses to RS's?

**Comment by robbat2** [comment]
Yes, leave the multicast cycle entirely alone.

In the worst case, immediately after we send a complete set of unicast RA to a single client, we have our regularly scheduled multicast, which for this window aligns to schedule ALL packets.

The overall objective here is that it should be feasible to INCREASE the multicast interval, but still get a near-immediate unicast response to RS, without flooding the network.

**Comment by robbat2** [comment]
Ok, this is going to be easiest to fix after RA-splitting is merged, because:
- it's going to conflict horribly otherwise
- in the case of no client list, we need an additional data structure of all _requesting_ nodes, so we can rate-limit each node to one RS per $INTERVAL (thinking of just using the MIN_DELAY_BETWEEN_RAS/MIN_DELAY_BETWEEN_RAS_MIPv6 values).

**Comment by spakka** [comment]
I'd like to add a +1 to this
Having a periodic multicast advertisement, but unicast responses to RS, would help us also with RFC7772 issues.
It would be good to be able to do this without having to configure a list of unicast clients as well.
We have many battery-powered devices coming and going from a large wireless lan, and every time one connects and sends an RS, all the others get their radio woken up by the multicast RA reply.

**Comment by robbat2** [comment]
Yes, this is still an issue; I started a patchset for it, but got sidetracked by other work.

The major issue was trying to decide on scheduling and tracking requesting nodes issues. The latter is a problem because it represents a potential DoS attack if we try to track all unicast clients (if they just spoof lots of different addresses, they can overwhelm radvd).

**Comment by robbat2** [comment]
@spakka can you read/test out PR #69?

**Comment by robbat2** [comment]
This PR represents the minimal fix. There are further improvements suggested in TODO for later, like deliberately deferring the unicast RA if we are getting close to the time of multicast RA; as well as forcing more multicast RAs when topology is changing.

**Comment by spakka** [comment]
Hi, thanks for the quick reply.
I've reviewed the patch and it looks in compliance with the RFC. I tried to test it but wasn't able to generate an RS with the SLLA option, so i moved the rfc7772_unicast_response test outside the option block, to make it apply to all RSs, and it works great. Only thing is that if a unicast response is destined for a link-local address and there is no existing ND entry on the router (and the SLLA option hasn't been provided in the RS), then it triggers an ND/NA from the router to resolve the link-local address so it can form the reply RA.
Anyway basically yea it works well as long as the sender sets the SLLA option.
Thanks!

**Comment by robbat2** [comment]
I can trigger it by having Linux restart an interface (libvirt instance of Ubuntu 14.04, patched quagga running on the bridge `virbr0`):
```
11:00:10.426779 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::5054:ff:feca:a591 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 52:54:00:ca:a5:91
	    0x0000:  5254 00ca a591
11:00:10.427420 IP6 (flowlabel 0xea489, hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::1 > fe80::5054:ff:feca:a591: [icmp6 sum ok] ICMP6, router advertisement, length 16
	hop limit 64, Flags [managed, other stateful], pref medium, router lifetime 1800s, reachable time 10000s, retrans time 5000s
```

In your prod environment, can you look for other RS and see how many of them do have SLLA set? If some OS doesn't properly set the SLLA option, it might be worthwhile to have a variant that doesn't require SLLA to be set, with a suitable warning that it would trigger multicast ND.

**Comment by spakka** [comment]
Not able to test the full prod environment as it is a meeting/conference network, and the next conference is in just over a month :)
But I tested with all the clients I have nearby, here are the results:
```
macOS 10.12.3 sierra - sends SLLA 2 times out of 4
iOS 10.2.1 (iPhone 5s) - no SLLA
Android 7.0 (sony xperia phone) - sends SLLA
Android 5.1 (nexus 7 tablet) - sends SLLA
Ubuntu 16.04.2 LTS w/ Network Manager, running 4.9 kernel (dell laptop) - no SLLA
Windows 10 (dell laptop) - no SLLA
```
So it isn't universally supported, and at least the iPhone is definitely in the category of small battery-based device (actually, all the above devices are battery-powered!) Not sure why macOS sometimes sets SLLA and sometimes doesn't.

Also, here are some observations from part of the IETF discussion of the draft that lead to RFC7772:
https://www.ietf.org/mail-archive/web/v6ops/current/msg22464.html

Note that you mention an option with a warning that it would trigger multicast ND from the router - on the router w/radvd that I tested, the NS was sent by the router to a solicited-node multicast address, targeting the link-local address it is trying to resolve. This is still better than sending to all-nodes, as the solicited-node multicast can be discarded by the device radio in non-targetted devices, without waking up the CPU. Whereas the all-nodes RA must wake up the CPU.

**Comment by reubenhwk** [comment]
I think I missed the SLLA requirement dropping question.  RADVD only set it
optionally.  Why is it in question right now?

**Comment by robbat2** [comment]
RFC7772 5.1.1 says to qualify for a unicast RA response, the RS needs to
1. not be from the unspecified address 
2. contain a SLLA option.

@spakka tested and found that lots of clients don't set an SLLA option in the RS (including Windows, NetworkManager & iOS mobile devices).

The language in the RFC does not say **MUST** contain an SLLA, so I propose to only have the unspecified address test.

**Comment by robbat2** [comment]
* The PR is implemented with both of the requirements at the moment, but could be changed to just unspecified easily.

**Comment by reubenhwk** [comment]
Ah.  Got it.  I'm ok with dropping the requirement.

**Comment by robbat2** [comment]
Updated to drop the requirement.
